crates/autopilot/src/domain/competition/auction_mechanism/mod.rs

```diff
@@ -0,0 +1,46 @@
+mod single_winner;
+
+pub use single_winner::SingleWinnerAuctionMechanism;
+use {
+    super::{Participant, Unranked},
+    crate::domain::Auction,
+    primitive_types::{H160, U256},
+};
+
+#[derive(Debug, thiserror::Error)]
+#[error("no winners found")]
+pub struct NoWinners;
+
+#[derive(Clone, Default, Debug)]
+pub struct ComputedScores {
+    // TODO: for now we specify a single winner as the database still expectes it
+    // After https://github.com/cowprotocol/services/issues/3350, it will no longer be necessary and we will be able to return only the vec or ReferenceScore
+    pub winner: H160,
+    pub winning_score: U256,
+    pub reference_scores: Vec<ReferenceScore>,
+}
+
+#[derive(Clone, Default, Debug)]
+pub struct ReferenceScore {
+    pub solver: H160,
+    pub reference_score: U256,
+}
+
+/// The following trait allows to implement custom auction mechanism logic
+/// for competitions.
+pub trait AuctionMechanism {
+    /// Filters out invalid or unfair solutions.
+    fn filter_solutions(
+        &self,
+        auction: &Auction,
+        solutions: &[Participant<Unranked>],
+    ) -> Vec<Participant<Unranked>>;
+
+    /// Selects the winners from a list of unranked solutions.
+    ///
+    /// Returns the list of solutions with the winners marked.
+    fn select_winners(&self, solutions: &[Participant<Unranked>]) -> Vec<Participant>;
+
+    /// Computes the scores of all provided solutions.
+    fn compute_scores(&self, solutions: &[Participant]) -> Result<ComputedScores, NoWinners>;
```

Comment:
I wonder how well the abstraction using this trait will actually work. IIUC we need to compute a winning score with each winner we pick (this I would expect in `select_winners()`) and we need to be able to compute reference scores for each winning solver. So I would expect something like `compute_reference_scores()` instead of `compute_scores()` with very different signatures.

Reply:
I kept the simplest function signatures for now to reduce the scope of this PR (which is to only abstract current logic).

But I expect the signatures to change to adapt to the full combinatorial implementation once it's in place. It will probably require to pass both winners and all solutions separately.